### PR TITLE
Update cobra dependency to v1.1.1

### PR DESCRIPTION
**Pull request by eddiezane**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

This PR bumps the vendored version of [cobra](https://github.com/spf13/cobra) to [v1.1.1](https://github.com/spf13/cobra/releases/tag/v1.1.1).

This version includes some changes to shell completion necessary for ongoing kubectl work.

**Which issue(s) this PR fixes**:

Refs https://github.com/kubernetes/kubectl/issues/882

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/cc @marckhouzam @brianpursley @knight42

### CODE REVIEW

**Comment by eddiezane** [comment]
Notes for reviewers: I've followed [this guide](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md) on updating. I'm not sure if the `gopkg.in/yaml.v2` are correct?

File: go.mod
Comment (liggitt) [inline go.mod:92]: running `hack/lint-dependencies.sh` will indicate this needs to be updated and vendor updated

File: go.mod
Comment (liggitt) [inline go.mod:112]: running `hack/lint-dependencies.sh` will indicate this needs to be updated and vendor updated

Comment (liggitt) [reply go.mod:112]: looks like this version changed the default line width, so we may need to explicitly set the previous default of 80 in our yaml serializers

Comment (eddiezane) [reply go.mod:111]: This is the related change https://github.com/go-yaml/yaml/pull/571 but I'm not sure where to start digging in on our end outside of seeing that the tests do.

**Comment by eddiezane** [comment]
@liggitt addressed the linting issues

File: go.mod
Comment (eddiezane) [reply go.mod:111]: And there are the failures. I'll dig in.

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/95571/pull-kubernetes-bazel-test/1316503504432402432

**Comment by laverya** [comment]
As the person who wrote the [relevant change](https://github.com/go-yaml/yaml/pull/571) in the underlying `gopkg.in/yaml.v2` library, all I can say is "I really should have expected this"

(That being said, "make the yaml marshaller not wrap lines that don't have a newline in the string" was the intended behavior change)

**Comment by eddiezane** [comment]
I dug in deep on the changes to gokpg.in/yaml.v2 and synced with the author of the change (@laverya). Trying to summarize.

Cobra [updated](https://github.com/spf13/cobra/pull/1153) off a vulnerable version of gopkg.in/yaml.v2 to v2.3.0.

This version of gopkg.in/yaml.v2 [changes](https://github.com/go-yaml/yaml/pull/571) the default line wrapping behavior from 80 characters to infinity. This value is **currently not configurable**. There was [another PR](https://github.com/go-yaml/yaml/pull/455) to the yaml library that added this behavior but the maintainer [opted](https://github.com/go-yaml/yaml/pull/455#issuecomment-624939185) to hold off on it hoping that folks wanted the infinity behavior.

Currently the only tests I've seen fail are [two in the conformance pipe](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/95571/pull-kubernetes-bazel-test/1316516890859802624) that are relying on the line wrapping.

3 options come to mind on proceeding.

1: Push for gopkg.in/yaml.v2 to make this value configurable. This would require Cobra updating and we would then need to create a new `yaml.Encoder` and set the line option everywhere we want this old behavior as well as educate and enforce to the rest of the project.

2: Leave the version of gopkg.in/yaml.v2 to pinned to v2.8.0. This would cause potential building and testing conflicts for other projects that import kubernetes since Cobra prefers the higher version.

3: Update the failing tests to not rely on the line wrapping behavior and hunt down any potential other conflicts. We'd also need to communicate to kdev about the new behavior.

My recommendation is to go with option 3. I've found a few issues that are looking for the new line wrapping behavior but are tied because of the version we use here. https://github.com/helm/helm/issues/8789, https://github.com/helm/helm/issues/7704

The other option is to not update Cobra but I imagine we'll need to deal with this at some point in the future.

Thoughts?

/cc @liggitt @pwittrock

**Comment by liggitt** [comment]
> 3: Update the failing tests to not rely on the line wrapping behavior and hunt down any potential other conflicts. We'd also need to communicate to kdev about the new behavior.

That will ripple far beyond k8s to every consumer... this behavior change is really unfortunate.

Examples of the sorts of data this will start failing tests on - https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/envtest/testdata/crdv1_updated/example_multiversion_crd1_one_more_version.yaml

**Comment by liggitt** [comment]
Given yaml.v2 is close to EOL, I'd actually push to revert the change and limit it to yaml.v3

**Comment by mattfarina** [comment]
Doesn't changing the default behavior like this violate SemVer? It should be an opt-in change from gopkg.in/yaml/v2.

**Comment by laverya** [comment]
> Doesn't changing the default behavior like this violate SemVer? It should be an opt-in change from gopkg.in/yaml/v2.

 Yuuuup. Or at least be backwards compatible by making it opt-in.

> Given yaml.v2 is close to EOL, I'd actually push to revert the change and limit it to yaml.v3. Control over line-wrapping is reasonable if it wanted to add that, but I'd expect the default behavior to remain consistent.

yaml.v3 doesn't have control over this either last I checked - it's the same 'infinite line length' behavior there too. (Also, there still isn't a v3.0.0 tag)

**Comment by marckhouzam** [comment]
From what I can see, Cobra upgraded from yaml.v2 2.2.2 to 2.3.0 to address a CVE reported by Kubernetes in https://github.com/kubernetes/kubernetes/issues/89535.  The fix for Kubernetes was to move to yaml.v2 2.2.8 as seen in https://github.com/kubernetes/kubernetes/pull/87467.

This means that Cobra didn't need to go to yaml.v2 2.3.0 but could have gone to 2.2.8.  Version 2.2.8 does not have the breaking change that is causing us problems now.

I'm not sure if this is a feasible solution, but if Cobra did an emergency release to revert back to yaml.v2 2.2.8, would that be enough to get rid of our current issues?

**Comment by liggitt** [comment]
> if Cobra did an emergency release to revert back to yaml.v2 2.2.8, would that be enough to get rid of our current issues?

That could be a stop-gap (would need to check if that's the only thing in the tree requiring yaml > 2.2.8), but it wouldn't hold long... it would only be a matter of time before some other dependency would rebump the yaml dep.

The real solution is for yaml.v2 to not change this default behavior

**Comment by laverya** [comment]
> The real solution is for yaml.v2 to not change this default behavior

This should really be an issue on the [yaml.v2 repo](https://github.com/go-yaml/yaml). From someone other than me, ideally...

**Comment by marckhouzam** [comment]
> That could be a stop-gap (would need to check if that's the only thing in the tree requiring yaml > 2.2.8)

Helm was already on 2.2.8 so I'm hopeful that version is acceptable.

> The real solution is for yaml.v2 to not change this default behavior

Agreed. If Cobra decides to do a revert it would only be in the mindset of allowing us to use the latest Cobra in the short-term.

For a long-term solution, I'll talk with helm maintainers about discussing the issue with the yaml.v2 project.

But @liggitt  you mention above that you suggest this change be part of yaml.v3.  What will `kubectl` do then? It will eventually have to move to yaml.v3 no?  I need to really understand the requirement before going to the yaml.v2 project with this issue.

Thanks

**Comment by liggitt** [comment]
> @liggitt you mention above that you suggest this change be part of yaml.v3. What will `kubectl` do then?

I'd like to see yaml.v3 default to backwards compatibility with yaml.v2, or at least give the option to be configured in a backwards compatible way

**Comment by jpmcb** [comment]
Cobra v1.1.1 is live: https://github.com/spf13/cobra/releases/tag/v1.1.1

and hopefully should provide a short term solution. It reverts to yaml.v2 2.2.8 per Marc's suggestion here: https://github.com/spf13/cobra/pull/1259

**Comment by eddiezane** [comment]
Thanks for the quick work, team!

I've updated this PR to use the newly cut version of cobra.

I'll open a convo upstream on go-yaml for now and we'll see where that goes.

/rename Update cobra dependency to v1.1.1

**Comment by eddiezane** [comment]
I've been emailing with the author of go-yaml (Gustavo) and he's been super kind and responsive. He's going to roll back the change for v2. Will keep tracking that and hold here for now.

/hold

**Comment by marckhouzam** [comment]
@eddiezane out of curiosity, with Cobra 1.1.1 using yalm.v2 2.2.8 I thought moving to Cobra 1.1.1 became independent of the yaml breaking change issue.  So I thought this PR could move forward.

**Comment by eddiezane** [comment]
@marckhouzam you're totally correct and if that's the preference here I can get the conflict resolved and we can merge. I got the sense that the go-yaml changes would be quick to land which would hopefully avoid any other transitive issues popping up. We can always bump twice though.

Perhaps we give until next week and if it's not cut we can proceed here?

Thoughts?

**Comment by marckhouzam** [comment]
> @marckhouzam you're totally correct and if that's the preference here I can get the conflict resolved and we can merge. I got the sense that the go-yaml changes would be quick to land which would hopefully avoid any other transitive issues popping up. We can always bump twice though.
> 
> Perhaps we give until next week and if it's not cut we can proceed here?

I don't expect Cobra to do another release in the short-term to take in any new yaml.v2 release.
That is why I thought going ahead with Cobra 1.1.1 now is probably best.

**Comment by fedebongio** [comment]
/remove-sig api-machinery

Comment: something is re adding removed SIGs, any clue anybody? Thanks!

**Comment by eddiezane** [comment]
@fedebongio probably because I've been force pushing like a jedi 😅. Sorry!

These failure is https://github.com/kubernetes/kubernetes/issues/95795

/test pull-kubernetes-bazel-build

**Comment by eddiezane** [comment]
@liggitt this is good to go if you're onboard.

TLDR: Cobra pinned good version of goyaml and cut a new release. Author of goyaml is going to revert the change at some point.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/triage accepted
/priority important-longterm

**Comment by marckhouzam** [comment]
PR #96087 moves all bash custom completions to Go, which paves the way to moving to native zsh completion.  However, Cobra 1.1.1 is required for that, so this PR would need to go in first.

**Comment by eddiezane** [comment]
@liggitt - rebased and ready to go

/remove-sig api-machinery

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by dims** [comment]
looks like this is ready to 🚢  @liggitt  (needs root OWNER approval)

**Comment by soltysh** [comment]
Good point Dims, worth
/hold cancel 
too ;)

**Comment by eddiezane** [comment]
@liggitt per Slack he's the list of new entries to `go.sum` with their license.

| Module | License |
| ------ | ------- |
| cloud.google.com/go/firestore | Apache-2.0 |
| github.com/armon/go-metrics | MIT |
| github.com/armon/go-radix | MIT |
| github.com/bketelsen/crypt | MIT |
| github.com/coreos/bbolt | MIT |
| github.com/coreos/etcd | Apache-2.0 |
| github.com/gorilla/websocket | BSD-2-Clause |
| github.com/hashicorp/consul/api | MPL-2.0 |
| github.com/hashicorp/consul/sdk | MPL-2.0 |
| github.com/hashicorp/errwrap | MPL-2.0 |
| github.com/hashicorp/go-cleanhttp | MPL-2.0 |
| github.com/hashicorp/go-immutable-radix | MPL-2.0 |
| github.com/hashicorp/go-msgpack | BSD-3-Clause |
| github.com/hashicorp/go-multierror | MPL-2.0 |
| github.com/hashicorp/go-rootcerts | MPL-2.0 |
| github.com/hashicorp/go-sockaddr | MPL-2.0 |
| github.com/hashicorp/go-uuid | MPL-2.0 |
| github.com/hashicorp/go.net | Golang Standard License (https://github.com/hashicorp/go.net/blob/master/LICENSE) |
| github.com/hashicorp/logutils | MPL-2.0 |
| github.com/hashicorp/mdns | MIT |
| github.com/hashicorp/memberlist | MPL-2.0 |
| github.com/hashicorp/serf | MPL-2.0 |
| github.com/mitchellh/cli | MPL-2.0 |
| github.com/mitchellh/go-testing-interface | MIT |
| github.com/mitchellh/gox | MPL-2.0 |
| github.com/mitchellh/iochan | MIT |
| github.com/pascaldekloe/goe | CC0 1.0 Universal (https://github.com/pascaldekloe/goe/blob/master/LICENSE https://creativecommons.org/publicdomain/zero/1.0/legalcode) |
| github.com/posener/complete | MIT |
| github.com/ryanuber/columnize | MIT |
| github.com/sean-/seed | MIT + Golang Standard License (https://github.com/sean-/seed/blob/master/LICENSE) |
| github.com/spf13/cobra | Apache-2.0 |
| github.com/spf13/viper | MIT |
| github.com/subosito/gotenv | MIT |
| gopkg.in/ini.v1 | Apache-2.0 |

**Comment by liggitt** [comment]
Thanks for gathering that. MPL and CC0 are not on the [allowed list of CNCF licenses](https://github.com/cncf/foundation/blob/master/allowed-third-party-license-policy.md#approved-licenses-for-allowlist), and I don't see the following in https://github.com/cncf/foundation/tree/master/license-exceptions:
* github.com/pascaldekloe/goe
* github.com/hashicorp/consul
* github.com/hashicorp/go-immutable-radix
* github.com/hashicorp/go-msgpack
* github.com/hashicorp/go-rootcerts
* github.com/hashicorp/go-sockaddr
* github.com/hashicorp/go-uuid
* github.com/hashicorp/logutils
* github.com/hashicorp/memberlist
* github.com/hashicorp/serf
* github.com/mitchellh/cli
* github.com/mitchellh/gox

**Comment by eddiezane** [comment]
This was a fun one.

open_SAKnExcG

It looks like all of these are pulled in via: github.com/spf13/cobra -> github.com/spf13/viper -> github.com/bketelsen/crypt -> github.com/hashicorp/consul/api

Seeing as this version of cobra is already in etcd I imagine we'll need to get these on the exception list. Do you know what the path for that is?

**Comment by liggitt** [comment]
I don't... @cblecker, do you know?

**Comment by liggitt** [comment]
cc @kubernetes/licensing

**Comment by dims** [comment]
@liggitt @eddiezane So what really matters is the LICENSES file that get added, not transitive, so for this PR those are

```
 create mode 100644 LICENSES/vendor/github.com/subosito/gotenv/LICENSE
 create mode 100644 LICENSES/vendor/gopkg.in/ini.v1/LICENSE
```

Given the both the packages have licenses that are covered by https://github.com/cncf/foundation/blob/master/allowed-third-party-license-policy.md#approved-licenses-for-allowlist
```
github.com/subosito/gotenv | MIT
gopkg.in/ini.v1            | Apache-2.0
```

We should be good.

Also, process-wise, the exception-list in CNCF lags what we release. basically we get to ask for exceptions as and when we need it. The CNCF GB approves/updates the exception list. We ping @swinslow when we need to ask questions. Typically @nikhita and myself have engaged Steve.

Thanks,
Dims

**Pull request merged**
